app/pages/EditBreakingNewsPage.jsx

```diff
@@ -0,0 +1,172 @@
+import React from 'react';
+import DatePicker from 'react-datepicker';
+import 'react-datepicker/dist/react-datepicker.css';
+import { RiAddBoxLine, RiDeleteBin5Line, RiSave3Line } from 'react-icons/ri';
+import { withRouter } from 'react-router-dom';
+
+import * as dataService from '../utils/DataService';
+import './EditBreakingNewsPage.scss';
+
+class EditBreakingNewsPage extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = { breakingNewsItems: [] };
+
+    this.onAddNew = this.onAddNew.bind(this);
+    this.onDelete = this.onDelete.bind(this);
+  }
+
+  async componentDidMount() {
+    const { news_articles } = await dataService.get('/api/news_articles');
+    const breakingNewsItems = news_articles.map(article => ({
+      ...article,
+      effective_date: this.formatDate(article.effective_date),
+      expiration_date: this.formatDate(article.expiration_date)
+    }));
+
+    this.setState({ breakingNewsItems });
+  }
+
+  formatDate(date) {
+    return date ? new Date(date) : null;
+  }
+
+  onFieldChange(item, { target }) {
+    const { value, dataset } = target;
+    const { field } = dataset;
+    const newItem = { ...item, [field]: value };
+
+    this.setState(prevState => ({
+      breakingNewsItems: prevState.breakingNewsItems
+        .map(oldItem => (oldItem.id === newItem.id ? newItem : oldItem)),
+    }));
+  }
+
+  onDatePickerChange(item, value, field) {
+    const newItem = { ...item, [field]: value };
+
+    this.setState(prevState => ({
+      breakingNewsItems: prevState.breakingNewsItems
+        .map(oldItem => (oldItem.id === newItem.id ? newItem : oldItem)),
+    }));
+  }
+
+  onSave(item) {
+    dataService.put(`/api/news_articles/${item.id}`, item);
+  }
+
+  async onDelete(item) {
+    await dataService.APIDelete(`/api/news_articles/${item.id}`);
+
+    this.setState(prevState => ({
+      breakingNewsItems: prevState.breakingNewsItems
+        .filter(oldItem => oldItem.id !== item.id),
+    }));
+  }
```

Comment:
All of these methods are concurrency/async unsafe due to accepting the entire `item` as an argument. React does not guarantee that a component will always be rerendered immediately after a state update (it happens to do this by chance on the ancient version of React we're still running, but the main React feature development over the past few years has been to explicitly not make this happen in order to support asynchronous rendering, which can greatly improve performance).

You have a data race in that the value of `item` comes from its value at _render_ time, as it is captured into the closures that you construct in the `onClick` properties, but this can drift from the _actual_ state of the component. For example, if you edit a field on a news item, it will trigger the `onFieldChange` method, but if you also hit the save button _before_ the component is rerendered, then the value of `item` passed to `onSave()` will be outdated compared to its corresponding `state` counterpart.

What you should really do is change each of these methods to accept [a handle](https://en.wikipedia.org/wiki/Handle_(computing)) to the news article, not a full copy of the news article data, either using the article's `id` property or its array index as the handle. Then you should only make modifications to the news articles from within the `setState()` call, referencing the previous state of the news article from the `prevState` argument to the function being passed to `setState()`. This ensures that you're always operating on the latest version of the item, even if the state were modified before the component is rerendered.